internal/controller/management_controller.go

```diff
@@ -92,101 +83,150 @@ func (r *ManagementReconciler) Update(ctx context.Context, management *hmc.Manag
+			updateComponentsStatus(statusAccumulator, component, nil, errMsg)
 			errs = errors.Join(errs, errors.New(errMsg))
+
 			continue
 		}
 
 		if component.Template != hmc.CoreHMCName {
 			if err := r.checkProviderStatus(ctx, component.Template); err != nil {
-				updateComponentsStatus(detectedComponents, &detectedProviders, detectedContracts, component.helmReleaseName, component.Template, template.Status.Providers, template.Status.CAPIContracts, err.Error())
+				updateComponentsStatus(statusAccumulator, component, nil, fmt.Sprintf("Failed to check provider status: %s", err))
 				errs = errors.Join(errs, err)
 				continue
 			}
 		}
 
-		updateComponentsStatus(detectedComponents, &detectedProviders, detectedContracts, component.helmReleaseName, component.Template, template.Status.Providers, template.Status.CAPIContracts, "")
+		updateComponentsStatus(statusAccumulator, component, template, "")
 	}
 
+	management.Status.AvailableProviders = statusAccumulator.providers
+	management.Status.CAPIContracts = statusAccumulator.compatibilityContracts
+	management.Status.Components = statusAccumulator.components
 	management.Status.ObservedGeneration = management.Generation
-	management.Status.AvailableProviders = detectedProviders
-	management.Status.CAPIContracts = detectedContracts
-	management.Status.Components = detectedComponents
 	management.Status.Release = management.Spec.Release
+
 	if err := r.Status().Update(ctx, management); err != nil {
-		errs = errors.Join(errs, fmt.Errorf("failed to update status for Management %s: %w",
-			management.Name, err))
+		errs = errors.Join(errs, fmt.Errorf("failed to update status for Management %s: %w", management.Name, err))
 	}
+
 	if errs != nil {
 		l.Error(errs, "Multiple errors during Management reconciliation")
 		return ctrl.Result{}, errs
 	}
+
 	return ctrl.Result{}, nil
 }
 
+func (r *ManagementReconciler) cleanupRemovedComponents(ctx context.Context, management *hmc.Management) error {
+	var (
+		errs error
+		l    = ctrl.LoggerFrom(ctx)
+	)
+
+	managedHelmReleases := new(fluxv2.HelmReleaseList)
+	if err := r.Client.List(ctx, managedHelmReleases,
+		client.MatchingLabels{hmc.HMCManagedLabelKey: hmc.HMCManagedLabelValue},
+		client.InNamespace(r.SystemNamespace), // all helmreleases are being installed only in the system namespace
+	); err != nil {
+		return fmt.Errorf("failed to list %s: %w", fluxv2.GroupVersion.WithKind(fluxv2.HelmReleaseKind), err)
+	}
+
+	for _, hr := range managedHelmReleases.Items {
+		componentName := hr.Name // providers(components) names map 1-1 to the helmreleases names
+
+		if componentName == hmc.CoreCAPIName ||
+			componentName == hmc.CoreHMCName ||
+			slices.ContainsFunc(management.Spec.Providers, func(newComp hmc.Provider) bool { return componentName == newComp.Name }) {
+			continue
+		}
+
+		l.Info("Found component to remove", "component_name", componentName)
+
+		if hr.Spec.ChartRef != nil {
```

Comment:
We should not remove helm charts here. They are associated with templates and will be garbage collected following template removal